### PR TITLE
adding ejbca micro-service

### PR DESCRIPTION
**Pull request by felipemuhamed**

### CODE REVIEW

File: pki/ejbca/.github/ISSUE_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+* **I'm submitting a ...**
```
Comment (mprevide) [inline pki/ejbca/.github/ISSUE_TEMPLATE.md]: Remove this file

File: pki/ejbca/.github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+* **Please check if the PR fulfills these requirements**
+- [ ] Tests for the changes have been added (for bug fixes / features)
```
Comment (mprevide) [inline pki/ejbca/.github/PULL_REQUEST_TEMPLATE.md]: Remove this file

File: pki/ejbca/.travis.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+language: node_js
```
Comment (mprevide) [inline pki/ejbca/.travis.yml]: remove this file

File: pki/ejbca/docker_compose.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+version: '2.1'
```
Comment (mprevide) [inline pki/ejbca/docker_compose.yaml]: I think This file is not necessary

File: pki/ejbca/travis/deploy-gh-pages.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
+git remote add -t gh-pages -f origin-gh-pages https://github.com/${TRAVIS_REPO_SLUG}
```
Comment (mprevide) [inline pki/ejbca/travis/deploy-gh-pages.sh]: Remove this file

File: pki/ejbca/travis/publish.sh [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash -ex
+if [ $TRAVIS_PULL_REQUEST == false ] ; then
```
Comment (mprevide) [inline pki/ejbca/travis/publish.sh]: Remove this file

File: pki/ejbca/api-docs/apiary.apib [hunk 7]
```diff
@@ -0,0 +1,490 @@
+FORMAT: 1A
+
+# EJBCA NODE API
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: EJBCA NODE is not intuitive, I'd change for something like REST Wrapper for EBJCA or EJBCA REST

File: pki/ejbca/api-docs/apiary.apib [hunk 8]
```diff
@@ -0,0 +1,490 @@
+FORMAT: 1A
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: Remove 'soap message:' from messages.

File: pki/ejbca/api-docs/apiary.apib [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
++ Response 200 (application/json)
+  
+        {
+          "status": 200,
+          "message": "ok"
+        }
+
+## CRL [/ca/{caname}/crl]
+
+### Create or update current CRL [PUT]
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: I didn't understand the idea of this method. Is this an action? If so, the method should be something like POST /ca/{caname}/generate_crl

File: pki/ejbca/api-docs/apiary.apib [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+
++ Response 201 (application/json)
+
+        {
+          "status": 201,
+          "message": "CRL Updated/Created"
+        }
+
+### get a CRL [GET /ca/{caname}/crl?{delta}]
+Use the command
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: It seems to be in the wrong place.

File: pki/ejbca/api-docs/apiary.apib [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+
++ Response 200 (application/json)
+
+        {
+          "CRL": "MIICQjCCASoCAQEwDQYJKoZIhvcNAQEFBQAwEzERMA8GA1UEAwwIUF..."
+        }
+
+# Group EJBCA user management
+## Create or edit user [/user]
+### Create or edit a user [POST]
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: You shouldn't use POST to edit. It is better to use PUT /user/\<username\>

File: pki/ejbca/api-docs/apiary.apib [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+          "moreinfo": "https://dojot.github.io/ejbca-rest/apiary_latest.html"
+        }
+
++ Response 200 (application/json)
+
+        {
+          "CRL": "MIICQjCCASoCAQEwDQYJKoZIhvcNAQEFBQAwEzERMA8GA1UEAwwIUF..."
+        }
+
+# Group EJBCA user management
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: EJBCA user management???

File: pki/ejbca/api-docs/apiary.apib [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+        {
+          "status": 200,
+          "message": "user created/edited with success."
+        }
+
+## User CRUD [/user/{username}]
+### get user data [GET]
+
++ Parameters
+  + username: IOTmidCA (required, string): user username
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: Is this right??

File: pki/ejbca/api-docs/apiary.apib [hunk 14]
```diff
@@ -0,0 +1,490 @@
...
+                  "subjectAltName": null,
+                  "subjectDN": "CN=device21",
+                  "tokenType": "USERGENERATED",
+                  "username": "device21"
+              }
+          ]
+        }
+
+
+### Revoke all cerificates from a user [DELETE /user/{username}?{reason,delete}]
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: This endpoint doesn't follow the good practices.

DELETE /user/{username} - remove the resource (user)
POST /user/{username}/revoke_all_certificates - revoke all certificates

File: pki/ejbca/api-docs/apiary.apib [hunk 15]
```diff
@@ -0,0 +1,490 @@
...
++ Response 200 (application/json)
+
+        {
+          "status": 200,
+          "message": "user deleted with success."
+        }
+
+
+## List user valid certificates [/user/{username}/find]
+### List user certs [GET /user/{username}/find?{valid}]
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: It's better to use a POST because it is an action

File: pki/ejbca/api-docs/apiary.apib [hunk 16]
```diff
@@ -0,0 +1,490 @@
...
+
++ Parameters
+  + username: IOTmidCA (required, string): user username
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: Is this right?

File: pki/ejbca/api-docs/apiary.apib [hunk 17]
```diff
@@ -0,0 +1,490 @@
...
+
++ Response 400 (application/json)
+
+        {
+          "code": 400,
+          "message": "soap message: query exception.",
+          "moreinfo": "https://dojot.github.io/ejbca-rest/apiary_latest.html"
+        }
+
++ Response 404 (application/json)
```
Comment (rascaraficci) [inline pki/ejbca/api-docs/apiary.apib]: 404 is an error code, it is better 200 with an empty array.

File: pki/ejbca/config_ra.sh [hunk 18]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+password=${EJBCA_PASS:-secret}
+
+#First we create a new end entity for soap_client
+/opt/primekey/bin/ejbca.sh ra addendentity --username soap_client --dn  \
+"CN=soap_client" --password ${password} --caname IOTmidCA \
+--type 1 --token P12
+
+#If the soap_client already exists (persisted) we renew his status
```
Comment (rascaraficci) [inline pki/ejbca/config_ra.sh]: ... we renew its status

File: pki/ejbca/package.json [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+  "bin": "index.js",
+  "scripts": {
+    "test": "jest tests/unit-tests/*.js --coverage",
+    "start": "node src/index.js",
+    "docs": "jsdoc -R ./README.md -p ./package.json -r -d ./docs --access all ./lib -c ./jsdoc.conf"
+  },
+  "keywords": [
+    "ejbca",
+    "dojot",
+    "security"
```
Comment (rascaraficci) [inline pki/ejbca/package.json]: change security by pki

File: pki/ejbca/package.json [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+  ],
+  "author": "Muhamed Avila",
+  "license": "GPL-3.0",
+  "dependencies": {
+    "@dojot/dojot-module-logger": "0.0.1-alpha.5",
+    "@jonaphael/dojot-module": "0.0.3",
+    "body-parser": "^1.19.0",
+    "express": "^4.17.1",
+    "express-validator": "^6.1.1",
+    "node-cache": "^4.2.1",
```
Comment (rascaraficci) [inline pki/ejbca/package.json]: Is it used?

File: pki/ejbca/package.json [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+  "keywords": [
+    "ejbca",
+    "dojot",
+    "security"
+  ],
+  "author": "Muhamed Avila",
+  "license": "GPL-3.0",
+  "dependencies": {
+    "@dojot/dojot-module-logger": "0.0.1-alpha.5",
+    "@jonaphael/dojot-module": "0.0.3",
```
Comment (rascaraficci) [inline pki/ejbca/package.json]: Is it used?

File: pki/ejbca/api-docs/apiary.apib [hunk 22]
```diff
@@ -0,0 +1,490 @@
...
+Be careful to properly copy the first and final line: the number of dashes (-)
+must be exactly as shown above.
+
+To view the contents of this certificate, execute the following command: 
+
+```bash
+openssl x509 -in ./Certificate.crt.pem -text -noout
+```
++ Parameters
+  + cacn: IOTmidCA (required, string): Certificate common name
```
[the remark this answers is not in the document]
Comment (felipemuhamed) [reply pki/ejbca/api-docs/apiary.apib]: this parameter is related with the CN(common name) of the certificate.

File: pki/ejbca/api-docs/apiary.apib [hunk 23]
```diff
@@ -0,0 +1,490 @@
...
+        }
+
+
+
+## Certificate status [/ca/{cacn}/certificate/{certserial}/status]
+
+### Check if a certificate is revoked [GET]
+
++ Parameters
+  + cacn: CN=IOTmidCA,O=EJBCA IOT,C=SE (required, string): Certificate distinguished name
```
[the remark this answers is not in the document]
Comment (felipemuhamed) [reply pki/ejbca/api-docs/apiary.apib]: This parameter is not related with the common name of the certificate. It related with the DN (distinguished name) of the certificate. I will change it name for better understanding of the api use.

File: pki/ejbca/api-docs/apiary.apib [hunk 24]
```diff
@@ -0,0 +1,490 @@
...
+
++ Response 404 (application/json)
+
+        {
+          "code": 404,
+          "message": "No certificate found",
+          "moreinfo": "https://dojot.github.io/ejbca-rest/apiary_latest.html"
+        }
+
++ Response 200 (application/json)
```
[the remark this answers is not in the document]
Comment (felipemuhamed) [reply pki/ejbca/api-docs/apiary.apib]: Has no difference.

File: pki/ejbca/api-docs/apiary.apib [hunk 25]
```diff
@@ -0,0 +1,490 @@
...
+  + cacn: CN=IOTmidCA,O=EJBCA IOT,C=SE (required, string): Certificate distinguished name
+  + certserial: 45A168F4EA978E7E (required, string): certificate serial number in hexadecimal
+  + reason: UNSPECIFIED (optional, enum) - one of REVOKATION_REASON enum values. default: UNSPECIFIED
+
+
++ Response 404 (application/json)
+
+        {
+          "code": 404,
+          "message": "Inexistent reason code",
```
[the remark this answers is not in the document]
Comment (felipemuhamed) [reply pki/ejbca/api-docs/apiary.apib]: It is, but if the user set a inexistent reason code, the ejbca will crash. If the user not set a reson code, the default is UNSPECIFIED.

**Pull request closed without merging**